### PR TITLE
Add specification for scheduler.yield() method

### PR DESCRIPTION
**Pull request by shaseley**
Implementation notes:
 - Refactor existing spec to share as much as possible between `yield()` and `postTask()`:
     - Extract out "scheduling state" struct to hold abort and priority sources and compute this based the options passed to `postTask()` and `yield()`
     - Change map data structures and related algorithms to additionally be keyed on "is continuation" boolean (the queues need to be separate since the effective priorities are different)
     - Unify `postTask()` and `yield()` priority with notion of an "effective priority", which is used for picking the next task to run (`yield()` effective priority for a given task priority is just higher than `postTask()` equivalent).
 - Implement inheritance by adding notion of "current scheduling state" to the event loop
     - Set before running `postTask()` callback and idle callbacks
     - Stored/restored in HostMakeJobCallback and HostCallJobCallback (i.e. propagate to microtasks from .then() call)

Note: things get a little awkward regarding continuation vs. task priority IDL since we can't yet inherit from enums (ContinuationPriority is TaskPriority + "inherit"). And we sometimes pass one type as another where they are compatible (i.e. same string values permitted).

***
<a href="https://pr-preview.s3.amazonaws.com/shaseley/scheduling-apis/pull/88.html" title="Last updated on May 15, 2024, 11:51 PM UTC (e586afb)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/scheduling-apis/88/11268d8...shaseley:e586afb.html" title="Last updated on May 15, 2024, 11:51 PM UTC (e586afb)">Diff</a>

### CODE REVIEW

File: spec/scheduling-tasks.md [hunk 1]
```diff
@@ -272,49 +362,105 @@ Issue: [=Run steps after a timeout=] doesn't necessarily account for suspension;
...
+  1. Set |handle|'s [=task handle/queue=] to the result of [=selecting the scheduler task queue=]
+     for |scheduler| given |state|'s [=scheduling state/priority source=] and true.
+  1. [=Schedule a task to invoke an algorithm=] for |scheduler| given |handle| and the following
+     steps:
+    1. Resolve |result|.
+  1. Return |result|.
+</div>
+
+<div algorithm>
+  To <dfn>compute the scheduling state from options</dfn> given a {{Scheduler}} object |scheduler|,
```
Comment (clelland) [inline spec/scheduling-tasks.md:395]: Overall, I find this algorithm hard to follow; I've written and deleted at least three comments where I thought there were errors already :)

Something structured like this might be clearer:

1. Let result be...
2. Let inheritedState be...
3. If abortOption is "inherit":
    * if inheritedState is not null, then set result’s abort source to inheritedState’s abort source.
4. Otherwise, set result’s abort source to abortOption.
5. If priorityOption is "inherit":
    * if inheritedState is not null, then set result’s priority source to inheritedState’s priority source.
6. Otherwise, if priorityOption is not null, then set result’s priority source to the result of creating a fixed priority unabortable task signal given priorityOption.
7. Otherwise, if abortOption is not null and implements the TaskSignal interface, then set result’s priority source to abortOption. _(although I would have expected this to reference result's abort source, rather than abortOption here)_
8. If result’s priority source is null, then set result’s priority source to the result of creating a fixed priority unabortable task signal given "user-visible".
9. Return result.

File: spec/scheduling-tasks.md [hunk 2]
```diff
@@ -272,49 +362,105 @@ Issue: [=Run steps after a timeout=] doesn't necessarily account for suspension;
...
-  1. If |priority| is null, |signal| is not null and [=implements=] the {{TaskSignal}} interface,
-     and |signal| [=TaskSignal/has fixed priority=], then set |priority| to |signal|'s
-     [=TaskSignal/priority=].
-  1. If |priority| is null and |signal| is not null and [=implements=] the {{TaskSignal}} interface,
-     then
+  To <dfn>schedule a yield continuation</dfn> for {{Scheduler}} |scheduler| given
+  {{SchedulerYieldOptions}} |options|:
+
+  1. Let |result| be [=a new promise=].
+  1. Let |abortOption| be |options|["{{SchedulerYieldOptions/signal}}"] if it [=map/exists=],
```
Comment (mmocny) [inline spec/scheduling-tasks.md:369]: Note: Should this be called abortOption given it could be used for more things?

Why not just `signalOption`, it would be less confusing even if it only supported abortcontroller.

File: spec/scheduling-tasks.md [hunk 3]
```diff
@@ -272,49 +362,105 @@ Issue: [=Run steps after a timeout=] doesn't necessarily account for suspension;
...
+  1. If |signal| is not null and it is [=AbortSignal/aborted=], then [=reject=] |result| with
+     |signal|'s [=AbortSignal/abort reason=] and return |result|.
+  1. Let |handle| be the result of [=creating a task handle=] given |result| and |signal|.
+  1. If |signal| is not null, then [=AbortSignal/add=] |handle|'s [=task handle/abort steps=] to
+     |signal|.
+  1. Set |handle|'s [=task handle/queue=] to the result of [=selecting the scheduler task queue=]
+     for |scheduler| given |state|'s [=scheduling state/priority source=] and true.
+  1. [=Schedule a task to invoke an algorithm=] for |scheduler| given |handle| and the following
+     steps:
+    1. Resolve |result|.
```
Comment (mmocny) [inline spec/scheduling-tasks.md:390]: Nice algo :D

File: spec/scheduling-tasks.md [hunk 4]
```diff
@@ -238,28 +324,32 @@ A <dfn>task handle</dfn> is a [=struct=] with the following [=struct/items=]:
...
+### Scheduling Tasks and Continuations ### {#sec-scheduler-alg-scheduling-tasks-and-continuations}
 
 <div algorithm>
   To <dfn>schedule a postTask task</dfn> for {{Scheduler}} |scheduler| given a
   {{SchedulerPostTaskCallback}} |callback| and {{SchedulerPostTaskOptions}} |options|:
 
   1. Let |result| be [=a new promise=].
-  1. Let |signal| be |options|["{{SchedulerPostTaskOptions/signal}}"] if
-     |options|["{{SchedulerPostTaskOptions/signal}}"] [=map/exists=], or otherwise null.
+  1. Let |state| be the result of [=computing the scheduling state from options=] given |scheduler|,
```
Comment (mmocny) [inline spec/scheduling-tasks.md:334]: Nit: "compute the scheduling state from options"

File: spec/scheduling-tasks.md [hunk 5]
```diff
@@ -272,49 +362,105 @@ Issue: [=Run steps after a timeout=] doesn't necessarily account for suspension;
...
+  1. Let |abortOption| be |options|["{{SchedulerYieldOptions/signal}}"] if it [=map/exists=],
+     otherwise null.
+  1. Let |priorityOption| be |options|["{{SchedulerPostTaskOptions/priority}}"] if it
+     [=map/exists=], otherwise null.
+  1. If |abortOption| is null and |priorityOption| is null, then set |abortOption| to
+     "{{SchedulerSignalInherit/inherit}}" and set |priorityOption| to
+     "{{SchedulerSignalInherit/inherit}}".
+  1. Otherwise if |abortOption| is "{{SchedulerSignalInherit/inherit}}" and |priorityOption| is null,
+     then set |priorityOption| to "{{SchedulerSignalInherit/inherit}}".
+  1. Let |state| be the result of [=computing the scheduling state from options=] given |scheduler|,
```
Comment (mmocny) [inline spec/scheduling-tasks.md:378]: "compute the scheduling state from options"

File: spec/controlling-tasks.md [hunk 6]
```diff
@@ -194,6 +194,15 @@ object |signal|, [=set/append=] |algorithm| to |signal|'s [=TaskSignal/priority
   1. Set |signal|'s [=TaskSignal/priority changing=] to false.
 </div>
 
+<div algorithm>
+  To <dfn>create a fixed priority unabortable task signal</dfn> given {{TaskPriority}} |priority|
+  and a |realm|.
```
[the remark this answers is not in the document]
Comment (shaseley) [reply spec/controlling-tasks.md:199]: Done.

File: spec/scheduling-tasks.md [hunk 7]
```diff
@@ -207,6 +279,20 @@ A <dfn>task handle</dfn> is a [=struct=] with the following [=struct/items=]:
   [=task/runnable=].
 </div>
 
+<div algorithm>
```
[the remark this answers is not in the document]
Comment (shaseley) [reply spec/scheduling-tasks.md:291]: Yeah that's probably easier to read. I don't know how it should be formatted, however (using defaults for now).

File: spec/scheduling-tasks.md (hunk 4 above)
Comment (shaseley) [reply spec/scheduling-tasks.md:347]: Any particular reason? It's common IIUC to use a different conjugation than the algorithm to make things more readable, and bikeshed handles the linking auto-magically (i.e. this links to "compute the ..." but is grammatically correct).

File: spec/scheduling-tasks.md [hunk 8]
```diff
@@ -238,28 +324,32 @@ A <dfn>task handle</dfn> is a [=struct=] with the following [=struct/items=]:
...
-     |options|["{{SchedulerPostTaskOptions/priority}}"] [=map/exists=], or otherwise null.
   1. Let |enqueueSteps| be the following steps:
     1. Set |handle|'s [=task handle/queue=] to the result of [=selecting the scheduler task queue=]
-       for |scheduler| given |signal| and |priority|.
-    1. [=Schedule a task to invoke a callback=] for |scheduler| given |callback|, |result|, and
-       |handle|.
+       for |scheduler| given |state|'s [=scheduling state/priority source=] and false.
+    1. [=Schedule a task to invoke an algorithm=] for |scheduler| given |handle| and the following
+       steps:
+      1. Let |event loop| be the [=relevant agent=] for |scheduler|'s [=agent/event loop=].
```
[the remark this answers is not in the document]
Comment (shaseley) [reply spec/scheduling-tasks.md]: Heh, I had it that way originally, but the definition reads like "The relevant agent for ...", so I matched it exactly -- but I see that apostrophe is ambiguous (it applies to [=relevant agent=] for |scheduler|).

I see a precedent of doing it the other way though, so I'll change it.

[the remark this answers is not in the document]
Comment (shaseley) [reply spec/scheduling-tasks.md]: Done.

File: spec/scheduling-tasks.md (hunk 5 above)
Comment (shaseley) [reply spec/scheduling-tasks.md:391]: Same question as above.

File: spec/scheduling-tasks.md [hunk 9]
```diff
@@ -272,49 +362,105 @@ Issue: [=Run steps after a timeout=] doesn't necessarily account for suspension;
...
-  To <dfn>schedule a task to invoke a callback</dfn> for {{Scheduler}} |scheduler| given a
-  {{SchedulerPostTaskCallback}} |callback|, a promise |result|, and a [=task handle=] |handle|:
+  To <dfn>schedule a task to invoke an algorithm</dfn> for {{Scheduler}} |scheduler| given a
+  [=task handle=] |handle| and an algorithm |steps|:
 
   1. Let |global| be the [=relevant global object=] for |scheduler|.
   1. Let |document| be |global|'s <a attribute for="Window">associated `Document`</a> if |global| is
      a {{Window}} object; otherwise null.
-  1. Let |event loop| be |scheduler|'s [=relevant agent's=] [=agent/event loop=].
+  1. Let |event loop| be the [=relevant agent=] for |scheduler|'s [=agent/event loop=].
```
[the remark this answers is not in the document]
Comment (shaseley) [reply spec/scheduling-tasks.md]: Done.

File: spec/scheduling-tasks.md (hunk 3 above)
Comment (shaseley) [reply spec/scheduling-tasks.md:403]: :)

File: spec/scheduling-tasks.md [hunk 10]
```diff
@@ -272,49 +362,105 @@ Issue: [=Run steps after a timeout=] doesn't necessarily account for suspension;
...
+  an {{AbortSignal}} object, "{{SchedulerSignalInherit/inherit}}", or null |abortOption| and a
+  {{TaskPriority}}, "{{SchedulerSignalInherit/inherit}}", or null |priorityOption|:
+
+  1. Let |result| be a new [=scheduling state=].
+  1. Let |inheritedState| be the [=relevant agent=] for |scheduler|'s [=agent/event loop=]'s
+     [=event loop/current scheduling state=].
+  1. If |abortOption| is "{{SchedulerSignalInherit/inherit}}", then set |result|'s
+     [=scheduling state/abort source=] to |inheritedState|'s [=scheduling state/abort source=] if
+     |inheritedState| is not null.
+  1. Otherwise |abortOption| is not "{{SchedulerSignalInherit/inherit}}"; set |result|'s
```
[the remark this answers is not in the document]
Comment (shaseley) [reply spec/scheduling-tasks.md]: Removed. FWIW I've seen this style in some WHATWG specs and thought it might help readability, but clearly not!

File: spec/scheduling-tasks.md [hunk 11]
```diff
@@ -272,49 +362,105 @@ Issue: [=Run steps after a timeout=] doesn't necessarily account for suspension;
...
+     [=scheduling state/abort source=] to |abortOption|.
+  1. If |priorityOption| is "{{SchedulerSignalInherit/inherit}}", then set |result|'s
+     [=scheduling state/priority source=] to |inheritedState|'s [=scheduling state/priority source=]
+     if |inheritedState| is not null.
+  1. Otherwise if |priorityOption| is not null, then set |result|'s [=scheduling state/priority
+     source=] to the result of [=creating a fixed priority unabortable task signal=] given
+     |priorityOption|.
+  1. Otherwise if |abortOption| is not null and [=implements=] the {{TaskSignal}} interface, then
```
[the remark this answers is not in the document]
Comment (shaseley) [reply spec/scheduling-tasks.md]: In this algorithm, |abortOption| _only_ applies to determining the abort component, in which case it's irrelevant.

If original signal should also be used for priority, then the caller will set the |priorityOption| to "inherit" (step 5 in the algo above). If the original signal was a TaskSignal, it gets set as both the abort and priority sources in the |inheritedState|, and it funnels into this check.

File: spec/scheduling-tasks.md (hunk 1 above)
Comment (shaseley) [reply spec/scheduling-tasks.md:408]: Heh, I want back and forth on this several times when drafting this. IIUC this was mostly rewriting 3 & 5? Just making sure I didn't miss anything (changed those in latest version).

Re: using `abortOption` in step 7 -- this handles the case where the arguments are of the form `{signal: someSignal}`, where there's no inheritance and the signal could be either a TaskSignal or plain AbortSignal. If it's a TaskSignal, then that priority gets used; otherwise, it defaults to "user-visible" in step 8.

File: spec/scheduling-tasks.md (hunk 2 above)
Comment (shaseley) [reply spec/scheduling-tasks.md:382]: Yeah good point. It's either change to signalOption or let priorityOption also be a signal. Changed to signalOption.

File: spec/scheduling-tasks.md [hunk 12]
```diff
@@ -24,18 +24,24 @@ updating secondary parts of the page. This is the default priority.
...
-<div algorithm>
-  {{TaskPriority}} |p1| is <dfn for="TaskPriority">less than</dfn>
-  {{TaskPriority}} |p2| if |p1| is less than |p2| in the
-  following total ordering:
-  "{{TaskPriority/background}}" < "{{TaskPriority/user-visible}}" < "{{TaskPriority/user-blocking}}"
-</div>
-
+tasks. Continuation priorities are slotted in just above their {{TaskPriority}} counterparts, e.g.
+a {{ContinuationPriority/user-visible}} continuation has a higher [=scheduler task queue/effective
+priority=] than a {{TaskPriority/user-visible}} task.
```
[the remark this answers is not in the document]
Comment (shaseley) [reply spec/scheduling-tasks.md:44]: One problem is that not all tasks will have an enqueue order, e.g. yielding in a `setTimeout()` since timers don't go through `scheduler`.

File: spec/scheduling-tasks.md (hunk 4 above)
Comment (mmocny) [reply spec/scheduling-tasks.md:347]: oh, cool.  I was just using find in doc.

File: spec/scheduling-tasks.md (hunk 1 above)
Comment (clelland) [reply spec/scheduling-tasks.md:408]: Thanks -- I missed that possibility

**Review by clelland — APPROVED** [review]

**Review by mmocny — APPROVED** [review]
